ext/node/polyfills/internal_binding/stream_wrap.ts

```diff
@@ -311,56 +311,61 @@ export class LibuvStreamWrap extends HandleWrap {
 
   /** Internal method for reading from the attached stream. */
   async #read() {
-    let buf = BUF;
-
-    let nread: number | null;
-    const ridBefore = this[kStreamBaseField]!.rid;
+    const isOwnedBuf = bufLocked;
+    let buf = bufLocked ? new Uint8Array(SUGGESTED_SIZE) : BUF;
+    bufLocked = true;
     try {
-      nread = await this[kStreamBaseField]!.read(buf);
-    } catch (e) {
-      // Try to read again if the underlying stream resource
-      // changed. This can happen during TLS upgrades (eg. STARTTLS)
-      if (ridBefore != this[kStreamBaseField]!.rid) {
-        return this.#read();
-      }
+      let nread: number | null;
+      const ridBefore = this[kStreamBaseField]!.rid;
+      try {
+        nread = await this[kStreamBaseField]!.read(buf);
+      } catch (e) {
+        // Try to read again if the underlying stream resource
+        // changed. This can happen during TLS upgrades (eg. STARTTLS)
+        if (ridBefore != this[kStreamBaseField]!.rid) {
+          return this.#read();
+        }
 
-      if (
-        e instanceof Deno.errors.Interrupted ||
-        e instanceof Deno.errors.BadResource
-      ) {
-        nread = codeMap.get("EOF")!;
-      } else if (
-        e instanceof Deno.errors.ConnectionReset ||
-        e instanceof Deno.errors.ConnectionAborted
-      ) {
-        nread = codeMap.get("ECONNRESET")!;
-      } else {
-        nread = codeMap.get("UNKNOWN")!;
-      }
+        if (
+          e instanceof Deno.errors.Interrupted ||
+          e instanceof Deno.errors.BadResource
+        ) {
+          nread = codeMap.get("EOF")!;
+        } else if (
+          e instanceof Deno.errors.ConnectionReset ||
+          e instanceof Deno.errors.ConnectionAborted
+        ) {
+          nread = codeMap.get("ECONNRESET")!;
+        } else {
+          nread = codeMap.get("UNKNOWN")!;
+        }
 
-      buf = new Uint8Array(0);
-    }
+        buf = new Uint8Array(0);
+      }
 
-    nread ??= codeMap.get("EOF")!;
+      nread ??= codeMap.get("EOF")!;
 
-    streamBaseState[kReadBytesOrError] = nread;
+      streamBaseState[kReadBytesOrError] = nread;
 
-    if (nread > 0) {
-      this.bytesRead += nread;
-    }
+      if (nread > 0) {
+        this.bytesRead += nread;
+      }
 
-    buf = buf.slice(0, nread);
+      buf = isOwnedBuf ? buf.subarray(0, nread) : buf.slice(0, nread);
 
-    streamBaseState[kArrayBufferOffset] = 0;
+      streamBaseState[kArrayBufferOffset] = 0;
 
-    try {
-      this.onread!(buf, nread);
-    } catch {
-      // swallow callback errors.
-    }
+      try {
+        this.onread!(buf, nread);
+      } catch {
+        // swallow callback errors.
+      }
 
-    if (nread >= 0 && this.#reading) {
-      this.#read();
+      if (nread >= 0 && this.#reading) {
+        this.#read();
+      }
+    } finally {
+      bufLocked = false;
```

Comment:
I am still encountering issue with concurrent read/writes using `ioredis` npm package.

I suspect this is the code that is causing the issue. Since `#read` is self-called, it can be that `bufLocked = false;` is set too late and `BUF` is still shared between concurrent calls.

Also, the following doesn't look ok:

```ts
let buf = bufLocked ? new Uint8Array(SUGGESTED_SIZE) : BUF;
...
buf = isOwnedBuf ? buf.subarray(0, nread) : buf.slice(0, nread);
...
if (nread >= 0 && this.#reading) {
   this.#read();
}
```

In case `nread >= 0` it should read from the same buf, but the code seems to read from a fresh `new Uint8Array(SUGGESTED_SIZE)`.

@mmastrac @bartlomieju

Reply:
Ahh. I think I see the problem. I unlocked the buffer regardless of whether we successfully obtained the lock. Follow-up PR coming ASAP.

Reply:
Fixed in https://github.com/denoland/deno/pull/20314